hudi-client/src/main/java/org/apache/hudi/execution/CopyOnWriteInsertHandler.java

```diff
@@ -0,0 +1,109 @@
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.execution.LazyInsertIterable.HoodieInsertValueGenResult;
+import org.apache.hudi.io.HoodieWriteHandle;
+import org.apache.hudi.io.WriteHandleFactory;
+import org.apache.hudi.table.HoodieTable;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Consumes stream of hoodie records from in-memory queue and writes to one or more create-handles.
+ */
+public class CopyOnWriteInsertHandler<T extends HoodieRecordPayload>
+    extends
+    BoundedInMemoryQueueConsumer<HoodieInsertValueGenResult<HoodieRecord>, List<WriteStatus>> {
+
+  protected HoodieWriteConfig config;
+  protected String instantTime;
+  protected HoodieTable<T> hoodieTable;
+  protected String idPrefix;
+  protected int numFilesWritten;
+  protected SparkTaskContextSupplier sparkTaskContextSupplier;
+  protected WriteHandleFactory<T> writeHandleFactory;
+
+  protected final List<WriteStatus> statuses = new ArrayList<>();
+  protected Map<String, HoodieWriteHandle> handles = new HashMap<>();
+
+  public CopyOnWriteInsertHandler(
+      HoodieWriteConfig config, String instantTime, HoodieTable<T> hoodieTable, String idPrefix,
+      SparkTaskContextSupplier sparkTaskContextSupplier, WriteHandleFactory<T> writeHandleFactory) {
+    this.config = config;
+    this.instantTime = instantTime;
+    this.hoodieTable = hoodieTable;
+    this.idPrefix = idPrefix;
+    this.numFilesWritten = 0;
+    this.sparkTaskContextSupplier = sparkTaskContextSupplier;
+    this.writeHandleFactory = writeHandleFactory;
+  }
+
+  @Override
+  public void consumeOneRecord(HoodieInsertValueGenResult<HoodieRecord> payload) {
+    final HoodieRecord insertPayload = payload.record;
+    String partitionPath = insertPayload.getPartitionPath();
+    HoodieWriteHandle handle = handles.get(partitionPath);
+    // lazily initialize the handle, for the first time
+    if (handle == null) {
+      handle = writeHandleFactory.create(
+          config, instantTime, hoodieTable, insertPayload.getPartitionPath(),
+          idPrefix, sparkTaskContextSupplier);
+      handles.put(partitionPath, handle);
+    }
+
+    if (handle.canWrite(payload.record)) {
+      // write the payload, if the handle has capacity
+      handle.write(insertPayload, payload.insertValue, payload.exception);
+    } else {
+      // handle is full.
+      statuses.add(handle.close());
+      // Need to handle the rejected payload & open new handle
+      handle = writeHandleFactory.create(
+          config, instantTime, hoodieTable, insertPayload.getPartitionPath(),
+          idPrefix, sparkTaskContextSupplier);
+      handles.put(partitionPath, handle);
+      handle.write(insertPayload, payload.insertValue,
+          payload.exception); // we should be able to write 1 payload.
+    }
+  }
+
+  @Override
+  public void finish() {
+    for (HoodieWriteHandle handle : handles.values()) {
```

Comment:
so this implies that `handle.close()` needs to be idempotent? iff we are closing the handle in L85 already, why close everything from the map again?

Reply:
okay makes sense. the  key thing seems to be that the handle replaced in the `handles` map, is just closed at the end.